controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,366 @@
+	WatchFilterValue string
+	ScopeFactory     scope.Factory
+	CaCertificates   []byte // PEM encoded ca certificates.
+
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=openstackfloatingippools,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=openstackfloatingippools/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ipam.cluster.x-k8s.io,resources=ipaddressclaims;ipaddressclaims/status,verbs=get;list;watch;update;create;delete
+// +kubebuilder:rbac:groups=ipam.cluster.x-k8s.io,resources=ipaddresses;ipaddresses/status,verbs=get;list;watch;create;update;delete
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *OpenStackFloatingIPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	_ = context.Background()
+	log := ctrl.LoggerFrom(ctx)
+
+	pool := &infrav1.OpenStackFloatingIPPool{}
+	if err := r.Client.Get(context.Background(), req.NamespacedName, pool); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if pool.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Add finalizer if it does not exist
+		if controllerutil.AddFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+			return ctrl.Result{Requeue: true}, r.Client.Update(context.Background(), pool)
+		}
+	} else {
+		// Handle deletion
+		return r.reconcileDelete(ctx, pool)
+	}
+
+	scope, err := r.ScopeFactory.NewClientScopeFromFloatingIPPool(ctx, r.Client, pool, r.CaCertificates, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.setIPStatuses(ctx, scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+
+		cluster, err := util.GetClusterFromMetadata(ctx, r.Client, claim.ObjectMeta)
+		if err != nil {
+			log.Info("Could not get cluster resource for IPAddressClaim")
+			return ctrl.Result{}, nil
+		}
+
+		infraCluster, err := r.getInfraCluster(ctx, cluster, &claim)
+		if err != nil {
+			return ctrl.Result{}, errors.New("error getting infra provider cluster")
+		} else if infraCluster == nil {
+			log.Info("infra cluster is not ready yet")
+			return ctrl.Result{}, nil
+		}
```

Comment:
We do this here because we call CAPO's GetOrCreateFloatingIP below, which takes an infraCluster argument, which it uses to get the cluster's external network. Doing this requires us to label the IPClaim with the cluster name.

I feel like this adds complexity and reduces the flexibility of this API. How about instead of all this we add a `FloatingIPNetwork` parameter to the pool spec. This would be a `NetworkFilter` which would be required to resolve to a single network, and whose resolved UUID we would write to the pool status.

The implications of this would be:
* A FloatingIPPool can only serve IPs from a single network
* That network must be specified on the pool

It would make things much easier to reason about, though. We would be able to get rid of all references to the cluster from the pool and claims.

Reply:
:+1: I agree that's probably a cleaner approach, that would make using a single pool for multiple clusters cleaner aswell or atleast reusing it for a new cluster.